### ISSUE TITLE
Features request

### ISSUE THREAD

**Issue by hippogamesunity** [hippogamesunity/PixelStudioHub#15]
Hi,

Hope you are doing well.

I have used Pixel Studio for some time now, I still love it, but I think it could be improved with some changes / features.

I have made some suggestion that I think could greatly improve my workflow:

When cloning a frame instead of creating a new frame at the end of the animation it would be very useful to create the new frame just after the cloned frame. It can be really annoying to « move backward » a bunch of frame one by one, especially when you have plenty of frame and you want to add several missing frames at a specific point of the animation. May be it could be an option in the settings menu.

I would do the same for the layers. When cloning a layer put the clone directly after the cloned layer instead of the top of all layers.

Possibility to hide a frame as you hide a layer. When I am animating I am not always sure that I should keep or note a frame, It would be nice to see my animation without deleting a frame by just hiding it (I have tried to change the delay to 0 but it doesn’t work)

Possibility to drag a bit offset when you are zoomed in. Sometimes it can be really tricky to touch pixels when they are near to the edge. I have to zoom a lot to edit the pixel which can be frustrating.

When using the flipping tool. Flip the whole layer if nothing is selected.

Rotate with a specific value (even if the result is not perfect it would be time saver)

Scale a specific area with a specific value or preset

Possibility to add more shortcut of custom tools. Now I am limited to one, as I have a pretty big phone screen I would love to add more than one

May be you have already thought about this and may be you are already working on it, in any case I wanted to share this with you. And I am really sorry if theses features already exist.

Thank you again for this amazing app.

Have a great weekend,

Simon

**Comment by Bearroneous** [comment]
Hi @hippogamesunity 

Sorry to add more onto your workload, but if I could throw a couple ideas onto the pile:

-More selection options than the existing rectangle, such as drawing a polygon by clicking where corners would be until the shape is closed by clicking the starting dot. Perhaps accessed by long-pressing the existing box selection tool.

-Cut, Copy & Paste toolbox that opens up when a selection tool is active.

-Color replacer brush, that will change a selected color into a 2nd selected color and ignore other colors. For when the user is testing colors or changing the color scheme of their drawing. (Sometimes I will rough out a picture in temporary colors that are easily distinguished from each other, before settling on a particular scheme, but currently I have to go back and flood-fill each pixel/area.
   -Building on this, being able to have 2 selected colors that can be changed between with a touch would be nice. A primary/secondary color tool, if you will.

-A button in the color picker to quickly erase the currently selected palette entry. I know we can zero-out the entry by changing it to 00000000, but that can take a while when blanking a handful of entries (while still wanting to keep the rest of the palette.)

-In the post above there is mention of allowing the canvas to be dragged further offscreen to easier edit the edges and such. As it is now, it seems fine for panning the canvas partially offscreen to get to the edges better, but the issue for me is it preventing me from zooming out beyond the borders of the canvas. Sometimes I like to work with a bit of white space or just to take a 'step back' and look at it with something of a border before its touching the tools/palette.

-Being able to start drawing while off canvas. This would be especially handy when using the line tool, as you could start it a few pixels outside of the canvas to ensure it touches the edge, instead of having to specifically start it in the very edge pixel. A bit of a deadzone border around the canvas for this would be nice to still allow panning to be done further away without drawing. A semi-transparent border line beyond the canvas to show this deadzone would be helpful. It only would need to be a few pixels wide (perhaps user defined in settings as well as turning it on/off?), just to give lee-way when beginning to use a tool that you want to come onto the canvas at the very edge.

-A UI Scale value in the settings to make the tool icons/palette bigger/smaller.

***********
A couple bugs/inconveniences I have noticed:
-The canvas pans around in the background when the color picker is open and I'm dragging the selection circle to a desired color. 

-I wouldn't mind just a tiny bit more of a movement deadzone for selecting from the palette instead of moving it. Sometimes when I try to select a color it thinks I am trying to scroll the palette even when there was minimal movement in the press, resulting in moving it only a few pixels and not selecting the color. (This is much more apparent when using a stylus to select the color, as the press area is so fine its easy to have it skitter just a couple pixels, which is enough for the app to think I'm trying to scroll instead of select. Finger touches are usually ok.)

-True Landscape Mode gets confused sometimes and displays the menus or canvas sideways. It seems to be related to the orientation I have the device in when I turn on True Landscape Mode. If I turn it on while the device is held in Portrait it works fine and the canvas stays in the correct direction. But if the device is held in landscape when I turn on True Landscape, the canvas bottom is move to the left side of the screen.
   -It would also be nice if it stayed at the current orientation when auto-rotate is turned off, instead of defaulting to portrait.

-The grid expands to encircle multiple pixels instead of staying 1:1 too early for me. I generally have a white background and a black 90% transparent grid, but unless I am really zoomed in, the grid encircles clusters of pixels instead of staying accurate. I understand why you have it expand when zoomed out, but the threshold seems to be too low before it steps apart. Could this be maybe a user defined threshold at certain zoom levels?
   -The grid also gets stuck sometimes. It doesn't seem to happen often, but sometimes when I zoom in and then later zoom back out the grid stays at the previous spacing (onscreen, not by canvas pixel count) instead of going back to the spacing it should have at the new zoom level. I am afraid I haven't been able to purposely recreate this bug. Restarting the app fixes it.

-This is probably too technical to be viable (I don't know anything about coding so I am sorry if this is a dumb question) but this app eats my phone's battery more than anything. I am guessing it is because it constantly refreshes the view at 60fps(30 in eco) but even in eco it chews quite a lot and the phone gets fairly warm (moreso than a dosbox emulator and some 3D games). Could the eco-mode's fps be user defined (I could get by with 10fps if it saved a bit of battery) or.. and this is the part I have no idea if its feasible, but, have it only update when something changes or the view is moved? Either way, I humbly request a look into more energy-saving options so I can poke pixels while out-and-about without tanking my battery.

-There is a lot of wasted space in True Landscape on either side of the tool area. Would it be possible to have this expand out sideways instead of vertically? (My plebeian thought is to have the tools be individual floating icons instead of a set sized toolbox, it would allow the whole width of the screen to be taken up before it made a new line, and more (custom) tools could be added or disabled dynamically.) I will attach a screenshot of my screen in landscape to show the extra unused space. I understand it won't be the same for all device's different resolutions.

*********
Thanks so much for hearing me out, I hope it wasnt too long, and like the poster above, I am very sorry if some of these features already exist and I just haven't explored the options enough (I spent most of writing this post while poking at the app just to make sure, but still..). Please know that these are only suggestions and I'm not trying to control your app or anything so selfish. I love the app and I'm glad you decided to make it in the first place! Thank you again... BRO!

![screenshot_20190218-225059_pixel studio](https://user-images.githubusercontent.com/47713617/52957393-ac22cd00-33e5-11e9-8d52-e678d9850355.jpg)

**Comment by hippogamesunity** [comment]
Yep, thanks =) By the way, you can replace a color using a Roller from Custom Tools. And if you disable True Landscape mode, you get side bars.

**Comment by Bearroneous** [comment]
Ohh so thats how the roller is meant to be used. Thanks! That will save a bit of time.

**Comment by newbe63** [comment]
Hi,
I'm from China, so my English was not good.If I say anything wrong,I'm sorry.
I like this app because I can draw pixel painting in ipad. It has a lot function, so I buy the PRO to use it. But only one function I can't find, it's pixel-perfect like Aseprite. I really need this function, it can save a lot of energy. So if you develop this function, I will recommend this app to my friends.
Thanks.

**Comment by hippogamesunity** [comment]
Hi! Do you mean Pixel Perfect Stroke from https://www.aseprite.org/?

**Comment by newbe63** [comment]
Yes! I need this function. Thanks!

**Comment by filcuk** [comment]
I've got a couple of suggestions too:

1: Option to resize the canvas
2: Option to import entire project into another as layers (not a picture)

**Comment by hippogamesunity** [comment]
Hello, use Fx menu to resize Project or Canvas.
